pipeline/authn/authenticator_oauth2_introspection.go

```diff
@@ -252,61 +254,71 @@ func (a *AuthenticatorOAuth2Introspection) Validate(config json.RawMessage) erro
 		return NewErrAuthenticatorNotEnabled(a)
 	}
 
-	_, err := a.Config(config)
+	_, _, err := a.Config(config)
 	return err
 }
 
-func (a *AuthenticatorOAuth2Introspection) Config(config json.RawMessage) (*AuthenticatorOAuth2IntrospectionConfiguration, error) {
+func (a *AuthenticatorOAuth2Introspection) Config(config json.RawMessage) (*AuthenticatorOAuth2IntrospectionConfiguration, *http.Client, error) {
 	var c AuthenticatorOAuth2IntrospectionConfiguration
 	if err := a.c.AuthenticatorConfig(a.GetID(), config, &c); err != nil {
-		return nil, NewErrAuthenticatorMisconfigured(a, err)
+		return nil, nil, NewErrAuthenticatorMisconfigured(a, err)
 	}
 
-	var rt http.RoundTripper
+	clientKey := fmt.Sprintf("%x", md5.Sum([]byte(config)))
+	a.mu.RLock()
+	client, ok := a.clientMap[clientKey]
+	a.mu.RUnlock()
 
-	if c.PreAuth != nil && c.PreAuth.Enabled {
-		var ep url.Values
+	if !ok {
```

Comment:
Should we add a test for this? I imagine a test which creates one instance for `AuthenticatorOAuth2Introspection`, uses three configs (one with preauth, one without preauth, one with another preauth) and then calls `Config` and expecnts the first config to always produce the same client (checking the pointer), the second never returns a client, and the third returns a different client. To make order random I would probably call config 1, 2, 3 - and then 2, 3, 1 to ensure this does not have side effects.

Reply:
I added a test, let me know what you think...